mobile/src/main/java/ch/epfl/sdp/mobile/ui/speech_recognition/SpeechRecognitionScreen.kt

```diff
@@ -0,0 +1,128 @@
+package ch.epfl.sdp.mobile.ui.speech_recognition
+
+import android.Manifest
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.speech.RecognitionListener
+import android.speech.RecognizerIntent
+import android.speech.SpeechRecognizer
+import androidx.compose.foundation.layout.Arrangement
+import androidx.compose.foundation.layout.Column
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.material.Icon
+import androidx.compose.material.OutlinedButton
+import androidx.compose.material.Text
+import androidx.compose.runtime.*
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.text.style.TextAlign
+import androidx.compose.ui.unit.dp
+import ch.epfl.sdp.mobile.ui.*
+import com.google.accompanist.permissions.ExperimentalPermissionsApi
+import com.google.accompanist.permissions.PermissionState
+import com.google.accompanist.permissions.rememberPermissionState
+import kotlin.coroutines.resume
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+const val Lang = "en-US"
+const val MaxResultsCount = 10
+
+@Composable
+@OptIn(ExperimentalPermissionsApi::class)
+fun SpeechRecognitionScreen(modifier: Modifier = Modifier) {
+
+  val context = LocalContext.current
+  var text by remember { mutableStateOf("---") }
+  val microPermissionState = rememberPermissionState(permission = Manifest.permission.RECORD_AUDIO)
+  var activeSpeech by remember { mutableStateOf(false) }
+  val scope = rememberCoroutineScope()
+
+  val microIcon = if (activeSpeech) PawniesIcons.GameMicOn else PawniesIcons.GameMicOff
+
+  Column(
+      verticalArrangement = Arrangement.spacedBy(32.dp, Alignment.CenterVertically),
+      horizontalAlignment = Alignment.CenterHorizontally,
+      modifier = modifier) {
+
+    // Speech Text
+    Text(text, textAlign = TextAlign.Center)
+
+    // Microphone Button
+    OutlinedButton(
+        shape = CircleShape,
+        onClick = {
+          askForPermission(microPermissionState)
+          activeSpeech = !activeSpeech && microPermissionState.hasPermission
+          scope.launch {
+            if (activeSpeech) {
+              text = "Listening..."
+              text = recognition(context).joinToString(separator = "\n")
+              activeSpeech = false
+            } else {
+              text = "---"
+            }
```

Comment:
You should use a `MutatorMutex` to make sure the previously launched recognition is properly cancelled.

Reply:
Added in `36fd0394853fb864c25e83883e8ce47bbd39ae79`